src/ess.jl

```diff
@@ -0,0 +1,351 @@
+            lastrange = lastrange .+ niter
+        else
+            # otherwise use views
+            s += dot(view(samples, firstrange, i), view(samples, lastrange, i)) / var[i]
+        end
+    end
+
+    return s / length(samples)
+end
+
+function mean_autocorr(k::Int, cache::FFTESSCache)
+    # check arguments
+    niter = cache.niter
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    A = cache.A
+    nchains = size(A, 2)
+    s = zero(real(eltype(A)))
+    @inbounds for i in 1:nchains
+        s += real(A[k + 1, i]) / real(A[1, i])
+    end
+
+    return s / nchains
+end
+
+function mean_autocorr(k::Int, cache::BDAESSCache)
+    # check arguments
+    samples = cache.samples
+    niter, nchains = size(samples)
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    var = cache.var
+    s = zero(eltype(var))
+    @inbounds for j in 1:nchains
+        sj = zero(s)
+        for i in 1:(niter - k)
+            sj += abs2(samples[i, j] - samples[k + i, j])
+        end
+        s += sj / var[j]
+    end
+
+    return 1 - s / (2 * length(samples))
+end
+
+"""
+    ess(chains::Chains; kwargs...)
+
+Estimate the effective sample size and the potential scale reduction.
+"""
+function ess(
+    chains::Chains;
+    sections = _default_sections(chains),
+    kwargs...
+)
+    # subset the chain
+    _chains = Chains(chains, _clean_sections(chains, sections))
+
+    # estimate the effective sample size and rhat
+    ess, rhat = ess_rhat(_chains.value.data; kwargs...)
+
+    # convert to namedtuple
+    nt = merge((parameters = names(_chains),), (ess = ess, rhat = rhat))
+
+    return ChainDataFrame("ESS", nt)
+end
+
+function ess_rhat(
+    chains::AbstractArray{<:Union{Missing,Real},3};
+    method::AbstractESSMethod = ESSMethod(),
+    maxlag::Int = 250
+)
+    # compute size of matrices (each chain is split!)
+    niter = size(chains, 1) ÷ 2
+    nparams = size(chains, 2)
+    nchains = 2 * size(chains, 3)
+    ntotal = niter * nchains
+
+    # do not compute estimates if there is only one sample or lag
+    maxlag = min(maxlag, niter - 1)
+    maxlag > 0 || return fill(missing, nparams), fill(missing, nparams)
+
+    # define caches for mean and variance
+    U = typeof(zero(eltype(chains)) / 1)
+    T = promote_type(eltype(chains), typeof(zero(eltype(chains)) / 1))
+    chain_mean = Array{T}(undef, 1, nchains)
+    chain_var = Array{T}(undef, nchains)
+    samples = Array{T}(undef, niter, nchains)
+
+    # compute correction factor
+    correctionfactor = niter / (niter - 1)
+
+    # define cache for the computation of the autocorrelation
+    esscache = build_cache(method, samples, chain_var)
+ 
+    # define output arrays
+    ess = Vector{T}(undef, nparams)
+    rhat = Vector{T}(undef, nparams)
+
+    # for each parameter
+    for (i, chains_slice) in enumerate(eachcol(chains))
```

Comment:
`eachcol` does not exist in Julia 1.0, I guess it's best if we add `Compat` (just add the package and load it).

Reply:
ie include it in the dependencies in `Project.toml`, and write `using Compat` at the beginning of `src/MCMCChains.jl`?

Reply:
You should set the compatibility bound to 2.2.0 (that's when `eachcol` was added to Compat).

Reply:
> ie include it in the dependencies in Project.toml, and write using Compat at the beginning of src/MCMCChains.jl?

Yes, exactly.

Reply:
Oops sorry, I didn't know that `eachcol` is apparently only defined for matrices, so you have to include both Compat (for Julia 1.0) and revert this line to
```suggestion
    for (i, chains_slice) in enumerate(eachslice(chains; dims = 2))
```

Reply:
:grimacing: